### PR TITLE
112 model

### PR DESCRIPTION
**Pull request by quinn-dougherty**
sorry, this is all over the place 
- Jason and I decided https://github.com/CodeForPhilly/chime/issues/112 wasn't really on the mark by recommending objects rather than functions
- even then, i largely didn't touch the chain of function calls that is the model because i thought they were suffficient  
- snuck in some housekeeping -
- - `# type: ignore` annotation to imports
- - took charting functions out of `presentation.py` into `charts.py` 
- - more things i've forgotten about sorry
- - moved `build_*_df` funcs from `utils.py` into `models.py` because they have mathematical content. 
- *mainly what i did:* changed signatures from `*args` to `p: Parameters` wherever possible.

### CODE REVIEW

File: src/app.py [hunk 1]
```diff
@@ -31,35 +28,15 @@
...
-    S=p.susceptible,
-    market_share=p.market_share,
-    recovery_days=p.recovery_days,
-    r_naught=p.r_naught,
-    doubling_time=p.doubling_time,
-    relative_contact_rate=p.relative_contact_rate,
-    r_t=p.r_t,
-    doubling_time_t=p.doubling_time_t,
-)
+display_header(st, p)
```
Comment (jlubken) [inline src/app.py:31]: Nice!

File: src/app.py [hunk 2]
```diff
@@ -68,30 +45,36 @@
 display_n_days_slider(st, p, DEFAULTS)
 
 # begin format data
-projection_admits = build_admissions_df(p.n_days, *p.dispositions)
-census_df = build_census_df(projection_admits, *p.lengths_of_stay)
+projection_admits = build_admissions_df(p=p) # p.n_days, *p.dispositions)
```
Comment (jlubken) [inline src/app.py]: projection_admits -> admissions_df

Comment (quinn-dougherty) [reply src/app.py]: what do you mean?

File: src/penn_chime/charts.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python
```
Comment (jlubken) [inline src/penn_chime/charts.py]: Remove the #!/usr/bin/python unless it is a script.

Comment (quinn-dougherty) [reply src/penn_chime/charts.py]: done, will be in next push

File: src/penn_chime/presentation.py [hunk 4]
```diff
@@ -252,12 +240,7 @@ def display_n_days_slider(st, p: Parameters, d: Constants):
 
 def show_more_info_about_this_tool(
     st,
-    recovery_days,
-    doubling_time,
-    r_naught,
-    relative_contact_rate,
-    doubling_time_t,
-    r_t,
+    parameters,
```
Comment (jlubken) [inline src/penn_chime/presentation.py:243]: Maybe just p instead of parameters

File: src/app.py (hunk 2 above)
Comment (jlubken) [reply src/app.py]: Both the census and the admissions are projections, I'd just simplify the variable to admissions_df for consistency.

**Comment by jlubken** [comment]
Very well done.

Neither Defaults nor Parameters are actually constants in the python sense, I wouldn't name the parameters PARAM, or DEFAULTS. I think most linters would complain.

No worries either way though.

**Comment by jlubken** [comment]
The test_sim_sir_df fails, likely because the signature changed.

File: src/penn_chime/charts.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+from typing import Optional
```
Comment (jlubken) [inline src/penn_chime/charts.py]: Lint says that this import isn't used anymore?

File: src/penn_chime/charts.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+def admitted_patients_chart(
+    alt,
+    census: pd.DataFrame,
+    parameters: Parameters,
+    as_date: bool = False
+) -> Chart:
+    """docstring"""
+
+    plot_projection_days = parameters.n_days - 10
```
Comment (jlubken) [inline src/penn_chime/charts.py:59]: Lint says that the variable plot_projection_days isn't used.

Comment (quinn-dougherty) [reply src/penn_chime/charts.py:59]: good catch -- that was something i had commented out while debugging that had forgotten to uncomment.

File: src/penn_chime/charts.py (hunk 5 above)
Comment (quinn-dougherty) [reply src/penn_chime/charts.py]: removed

File: src/penn_chime/charts.py (hunk 6 above)
Comment (quinn-dougherty) [reply src/penn_chime/charts.py:59]: fixed.

File: src/app.py (hunk 2 above)
Comment (quinn-dougherty) [reply src/app.py]: i see. i agree with that. will resolve this conversation as soon as i push

**Comment by jlubken** [comment]
I missed the fact that PARAM and DEFAULTS were constants in the test suite. Sorry about that.

**Comment by jlubken** [comment]
To prevent circular type annotation complains use:

`from __future__ import annotations` at the top of the file before any other imports.

File: src/penn_chime/presentation.py (hunk 4 above)
Comment (quinn-dougherty) [reply src/penn_chime/presentation.py:243]: this is actually a preference of mine. part of it is that i worked with a linter for a long time that disabled one letter variables. the other part of it is that i like calling on keywords `parameters=p` better than `p=p`. (I know it's not uniform across the PR, which is my mistake). but above all, the i think entries like `doubling_time=parameters.doubling_time` are easier to read than `doubling_time=p.doubling_time`.

i'm not _married_ to it. do you think there's a big add from keeping it uniform w the one-letter variable that's already all over the codebase?

**Comment by quinn-dougherty** [comment]
> I missed the fact that PARAM and DEFAULTS were constants in the test suite. Sorry about that. 

this is off topic, but i consider our `Constants` instance constant with respect to runtime, because the machine doesn't edit `settings.py` the human does (preferably before they run `streamlit run src/app.py`)

File: src/penn_chime/presentation.py (hunk 4 above)
Comment (jlubken) [reply src/penn_chime/presentation.py:243]: You are right about the linter. Why couldn't I thought of a shorter name for the thing...

This is just fine.

I was probably thinking about interpolating f strings directly. That would get quite long with the long variable name.

**Review by jlubken — APPROVED** [review]

**Pull request merged**